src/ui/src/core_components/content/CoreAnnotatedText.vue

```diff
@@ -0,0 +1,339 @@
+import injectionKeys from "../injectionKeys";
+import { inject } from "vue";
+import chroma, { Color } from "chroma-js";
+
+const fields = inject(injectionKeys.evaluatedFields);
+
+const COLOR_STEPS = [
+	{ h: -78, s: -34, l: 16 },
+	{ h: -61, s: -37, l: 16 },
+	{ h: -2, s: 0, l: 24 },
+	{ h: -12, s: 0, l: 29 },
+	{ h: 28, s: -20, l: 24 },
+	{ h: -61, s: -95, l: 25 },
+	{ h: -173, s: 0, l: 16 },
+	{ h: -228, s: 0, l: 22 },
+	{ h: 69, s: 0, l: 25 },
+	{ h: 70, s: 0, l: 29 },
+];
+
+let currentSteps = [...COLOR_STEPS];
+let subjectColorCache = {};
+let lastSeed = fields.seed.value;
+
+function generateColorCss(
+	baseColor: Color,
+	colorData: { h: number; s: number; l: number },
+) {
+	let genColor = baseColor
+		.set(
+			"hsl.h",
+			`${Math.sign(colorData.h) == -1 ? "-" : "+"}${Math.abs(colorData.h)}`,
+		)
+		.set(
+			"hsl.s",
+			`${Math.sign(colorData.s) == -1 ? "-" : "+"}${Math.abs(colorData.s / 100.0)}`,
+		)
+		.set(
+			"hsl.l",
+			`${Math.sign(colorData.l) == -1 ? "-" : "+"}${Math.abs(colorData.l / 100.0)}`,
+		);
+
+	return genColor.css();
+}
+
+function calculateColorStep(s: string, stepsLength = COLOR_STEPS.length) {
+	let hash = (fields.seed.value * 52673) & 0xffffffff;
+	for (let i = 0; i < s.length; i++) {
+		hash = ((i + 1) * s.charCodeAt(i)) ^ (hash & 0xffffffff);
+	}
+	const step = Math.abs(hash) % stepsLength;
+
+	return step;
+}
+
+function generateColor(s: string) {
+	if (fields.rotateHue.value == "no") {
+		return fields.referenceColor.value;
+	}
+
+	const baseColor = chroma(fields.referenceColor.value);
+
+	if (lastSeed !== fields.seed.value) {
+		currentSteps = [...COLOR_STEPS];
+		subjectColorCache = {};
+		lastSeed = fields.seed.value;
+	}
+
+	if (subjectColorCache[s]) {
+		return generateColorCss(baseColor, subjectColorCache[s]);
+	}
+
+	// If we run out of colors, reset the list
+	if (currentSteps.length === 0) {
+		currentSteps = [...COLOR_STEPS];
+	}
+
+	const colorStep = calculateColorStep(s, currentSteps.length);
+	const colorData = currentSteps[colorStep];
+
+	subjectColorCache[s] = colorData;
+	currentSteps.splice(colorStep, 1);
+
+	return generateColorCss(baseColor, colorData);
+}
+
+function copyText(arr: string[]) {
+	const text = arr.reduce((acc, val) => {
+		if (typeof val === "string") {
+			return acc + val;
+		}
+
+		return acc + val[0];
+	}, "");
+
+	copyToClipboard({ text });
+}
+
+function copyJSON(arr: string[]) {
+	try {
+		copyToClipboard({ text: JSON.stringify(arr) });
+	} catch (e) {
+		copyToClipboard({ text: arr.join("") });
+	}
+}
+
+function setClipboardData<T = unknown>(
+	source: T & { clipboardData: DataTransfer | null | undefined },
+	{ text, html }: { text?: string; html?: string },
+): void {
+	if (text) {
+		source.clipboardData?.setData("text/plain", text);
+		source.clipboardData?.setData("Text", text); // IE mimetype
+	}
+
+	if (html) {
+		source.clipboardData?.setData("text/html", html);
+	}
+}
+
+function copyToClipboard({
```

Comment:
I believe these two copy functions should travel someplace else but I have no understanding where to.